Structure_Engine/Query/ReinforcementLayout.cs

```diff
@@ -0,0 +1,207 @@
+        [Input("section", "The concrete section to return all points from.")]
+        [Input("position", "Position along the section to extract reinforcement. A negative value will return all reinforcement.")]
+        [Output("points", "The positions of the LongitudinalReinforcement.")]
+        public static List<Point> ReinforcementLayout(ConcreteSection section, double position = -1)
+        {
+            //Extract Longitudinal reinforcement
+            List<LongitudinalReinforcement> longReif = section.LongitudinalReinforcement();
+
+            //No longitudinal reinforcement available
+            if (longReif.Count == 0)
+                return new List<Point>();
+
+            List<ICurve> outerProfileEdges;
+            List<ICurve> innerProfileEdges;
+
+            ExtractInnerAndOuterEdges(section, out outerProfileEdges, out innerProfileEdges);
+
+            //Need at least one external edge curve
+            if (outerProfileEdges.Count == 0)
+                return new List<Point>();
+
+            //TODO: include stirups for offset distance
+            double stirupOffset = 0;
+            double cover = section.MinimumCover + stirupOffset;
+
+            List<Point> rebarPoints = new List<Point>();
+
+            foreach (LongitudinalReinforcement reif in longReif)
+            {
+                if (position < 0 || (reif.StartLocation <= position && reif.EndLocation >= position))
+                {
+                    rebarPoints.AddRange(ReinforcementLayout(reif, cover, outerProfileEdges, innerProfileEdges));
+                }
+            }
+
+            return rebarPoints;
+        }
+
+        /***************************************************/
+
+        [Description("Gets the LongitudinalReinforcement positions as a list of points, based on inner and outer profile edges.")]
+        [Input("reinforcement", "The LongitudinalReinforcement to extract the points from.")]
+        [Input("cover", "Any additional spacing to be added in relation to the edge curves, such as minimum cover and stirups. This will be added on top of half the rebar Diameter.", typeof(Length))]
+        [Input("outerProfileEdges", "The outer profile edges of the ConcreteSection to be populated.")]
+        [Input("innerProfileEdges", "The inner profile edges, or openings, of the ConcreteSection to be populated.")]
+        [Output("points", "The positions of the LongitudinalReinforcement.")]
+        public static List<Point> ReinforcementLayout(this LongitudinalReinforcement reinforcement, double cover, List<ICurve> outerProfileEdges, List<ICurve> innerProfileEdges = null)
+        {
+            innerProfileEdges = innerProfileEdges ?? new List<ICurve>();
+            double offset = cover + reinforcement.Diameter / 2;
+            IEnumerable<ICurve> outerCurves = outerProfileEdges.Select(x => x.IOffset(offset, -Vector.ZAxis));
+            IEnumerable<ICurve> innerCurves = innerProfileEdges.Select(x => x.IOffset(offset, Vector.ZAxis));
```

Comment:
This can offset outwards depending on curve orientation if one creates a free form profile.
![image](https://user-images.githubusercontent.com/53530468/82050514-2953f500-96b0-11ea-9021-5c15c98fc853.png)